render/gles2/surface.c

```diff
@@ -34,6 +34,35 @@ static bool gles2_surface_attach_pixels(struct wlr_surface_state *surface,
 	return true;
 }
 
+static bool gles2_surface_attach_shm(struct wlr_surface_state *surface,
+		uint32_t format, struct wl_shm_buffer *buffer) {
+	const struct pixel_format *fmt = gl_format_for_wl_format(format);
+	if (!fmt || !fmt->gl_format) {
+		wlr_log(L_ERROR, "No supported pixel format for this surface");
+		return false;
+	}
+	wl_shm_buffer_begin_access(buffer);
+	uint8_t *pixels = wl_shm_buffer_get_data(buffer);
+	int width = wl_shm_buffer_get_width(buffer);
+	int height = wl_shm_buffer_get_height(buffer);
+	int stride = wl_shm_buffer_get_stride(buffer);
+	surface->wlr_surface->width = width;
+	surface->wlr_surface->height = height;
+	surface->wlr_surface->format = format;
+	surface->pixel_format = fmt;
+
+	GL_CALL(glActiveTexture(GL_TEXTURE0));
```

Comment:
This is a really good explanation of texture units that I read: https://open.gl/textures#Textureunits